### PR TITLE
Make sure new record are not pushed twice when parent is saved before.

### PR DESCRIPTION
**Pull request by olivierchatry**

### CODE REVIEW

**Comment by steffenbrem** [comment]
@olivierchatry Does it also update the record again when new data is returned by the server (such as the id for example)?

**Comment by olivierchatry** [comment]
yes, it should.

**Comment by steffenbrem** [comment]
Nice, would be cool to see this in core :+1:

File: addon/-private/system/relationships/state/relationship.js [hunk 1]
```diff
@@ -173,7 +173,9 @@ Relationship.prototype = {
     //TODO(Igor) make this less abysmally slow
     this.members = this.canonicalMembers.copy();
     for (i=0; i<newRecords.length; i++) {
-      this.members.add(newRecords[i]);
+      if (!this.members.has(newRecords[i])) {
+        this.members.add(newRecords[i]);
+      }
```
Comment (bmac) [inline addon/-private/system/relationships/state/relationship.js]: @olivierchatry did you run into an issue that required this if check? Under the hood `members` does a [similar check](https://github.com/emberjs/ember.js/blob/master/packages/ember-metal/lib/map.js#L113) which should be preventing the record from being added twice.

File: tests/integration/relationships/has-many-test.js [hunk 2]
```diff
@@ -414,6 +414,64 @@ test("A hasMany updated link should not remove new children when the parent reco
...
+                data: [
+                  {
+                    id: 5,
+                    type: "comment"
+                  }
+                ]
+              }
+            }
+          }
+        });
```
Comment (bmac) [inline tests/integration/relationships/has-many-test.js]: @olivierchatry is the `store.push` here important to this test or can we move the info about the `comments` relationship into the return form the stubbed `createRecord` method?

Comment (olivierchatry) [inline tests/integration/relationships/has-many-test.js]: No it is important it showcase the bug.

File: addon/-private/system/relationships/state/relationship.js (hunk 1 above)
Comment (olivierchatry) [inline addon/-private/system/relationships/state/relationship.js]: Yes, using firebase, record get duplicate, see test.

Comment (bmac) [inline addon/-private/system/relationships/state/relationship.js]: When I remove this `if` check all of the tests still pass.

File: tests/integration/relationships/has-many-test.js (hunk 2 above)
Comment (bmac) [inline tests/integration/relationships/has-many-test.js]: Seems good. :+1:

Comment (olivierchatry) [inline tests/integration/relationships/has-many-test.js]: The fixes is in two seperate files, but maybe we only need to do it in one ?

**Comment by olivierchatry** [comment]
So, what do I need to do to make it OK ?

**Comment by bmac** [comment]
3 things:
1. I'd like to get @igorT to review this pr.
2. I think the change in `addon/-private/system/relationships/state/relationship.js` can be reverted.
3. Would you mind squashing this pr down to 1 commit and prefixing the commit message with `[BUGFIX beta]` so it can be pulled into the 2.4 beta branch.

**Comment by olivierchatry** [comment]
I will revert the relationship one as it is redundant. How do you want me
to squash down the PR ? create a new one ?

**Comment by bmac** [comment]
@olivierchatry you can rebase your branch locally and then force push it to github and it will update this pr.

Here is a good short tutorial on how to perform a rebase locally: https://egghead.io/lessons/javascript-how-to-squash-multiple-git-commits

**Comment by olivierchatry** [comment]
Am turning a bit crazy, wondering if it would not be easier to create a new branch and a new pull request from it.

**Comment by steffenbrem** [comment]
@olivierchatry Do the following commands to squash your stuff:

```bash
git rebase -i HEAD~6
```

It will open vim or something equalvant. Then replace `pick` with `s` on the last `5` commits and save + exit the file.

Your history is now been altered to one single commit, now run `git push -f` to force push your changes to the remote.

**Comment by olivierchatry** [comment]
stupid me, was missing the -f on the push ... thanks a lot !

**Comment by sirvine** [comment]
I notice this isn't in 2.5.0-beta1. Should we expect this to appear in a release soon?

**Comment by bmac** [comment]
I'll make sure this gets included in the `2.5.0-beta.2` release.

**Pull request merged**